core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/CompositeRuinMoveSelector.java

```diff
@@ -0,0 +1,266 @@
+            new HashMap<>();
+    private final boolean chained;
+
+    public CompositeRuinMoveSelector(EntitySelector<Solution_> leftEntitySelector,
+            EntitySelector<Solution_> rightEntitySelector,
+            List<GenuineVariableDescriptor<Solution_>> variableDescriptorList,
+            Collection<ShadowVariableDescriptor<Solution_>> shadowVariableDescriptors, Integer percentageToBeRuined) {
+        assert 0 < percentageToBeRuined && percentageToBeRuined <= 100
+                : "CompositeRuinMoveSelector: percentage to ruin the solution must be greater than 0 and smaller or equal to 100%";
+        this.leftEntitySelector = leftEntitySelector;
+        this.rightEntitySelector = rightEntitySelector;
+        this.variableDescriptorList = variableDescriptorList;
+        this.shadowVariableDescriptors = shadowVariableDescriptors;
+        this.percentageToBeRuined = percentageToBeRuined;
+        this.chained = this.variableDescriptorList.stream().anyMatch(GenuineVariableDescriptor::isChained);
+
+        phaseLifecycleSupport.addEventListener(leftEntitySelector);
+        if (leftEntitySelector != rightEntitySelector) {
+            phaseLifecycleSupport.addEventListener(rightEntitySelector);
+        }
+    }
+
+    // ************************************************************************
+    // Worker methods
+    // ************************************************************************
+
+    @Override
+    public boolean isCountable() {
+        return leftEntitySelector.isCountable() && rightEntitySelector.isCountable();
+    }
+
+    @Override
+    public boolean isNeverEnding() {
+        return true;
+    }
+
+    @Override
+    public long getSize() {
+        return percentageToBeRuined * leftEntitySelector.getSize() / 100;
+    }
+
+    @Override
+    public void solvingStarted(SolverScope<Solution_> solverScope) {
+        super.solvingStarted(solverScope);
+        if (chained) {
+            SupplyManager supplyManager = solverScope.getScoreDirector().getSupplyManager();
+            for (GenuineVariableDescriptor<Solution_> variableDescriptor : variableDescriptorList) {
+                if (variableDescriptor.isChained()) {
+                    variableDescriptorToInverseVariableSupply.put(variableDescriptor,
+                            supplyManager.demand(new SingletonInverseVariableDemand<>(variableDescriptor)));
+                }
+            }
+        }
+    }
+
+    @Override
+    public void solvingEnded(SolverScope<Solution_> solverScope) {
+        super.solvingEnded(solverScope);
+        if (chained) {
+            variableDescriptorToInverseVariableSupply.clear();
+        }
+    }
+
+    @Override
+    public Iterator<Move<Solution_>> iterator() {
+        return new RuinMoveIterator();
+    }
+
+    private class RuinMoveIterator extends UpcomingSelectionIterator<Move<Solution_>> {
```

Comment:
This iterator is extremely complex. 150 lines of code. Please break it down into smaller logical units, such as methods, with descriptive names and comments. Also, I'd expect a lot of test coverage for such a complex piece of code.

Reply:
done, but requires some more testing